File: app/src/main/java/net/myacxy/squinch/helpers/tracking/builder/PropertyLogBuilder.java
package net.myacxy.squinch.helpers.tracking.builder;

import net.myacxy.squinch.helpers.tracking.TrackingHelper;

import java.util.HashMap;
import java.util.Map;

public abstract class PropertyLogBuilder<T extends PropertyLogBuilder> extends BaseLogBuilder<T> {

    protected Map<String, String> properties = new HashMap<>();

    public PropertyLogBuilder(TrackingHelper trackingHelper) {
        super(trackingHelper);
    }

    public T addProperty(String key, String value) {
        properties.put(key, value);
        return self();
    }
}

File: app/src/main/java/net/myacxy/squinch/helpers/tracking/builder/PropertyLogBuilder.kt
package net.myacxy.squinch.helpers.tracking.builder

import net.myacxy.squinch.helpers.tracking.TrackingHelper
import java.util.*

abstract class PropertyLogBuilder<out T : PropertyLogBuilder<T>>(trackingHelper: TrackingHelper) : BaseLogBuilder<T>(trackingHelper) {

    protected val booleans: MutableMap<String, Boolean> = HashMap()
    protected val ints: MutableMap<String, Int> = HashMap()
    protected val strings: MutableMap<String, String> = HashMap()

    fun addProperty(key: String, value: Boolean): T {
        booleans[key] = value
        return self()
    }

    fun addProperty(key: String, value: Int): T {
        ints[key] = value
        return self()
    }

    fun addProperty(key: String, value: String): T {
        strings[key] = value
        return self()
    }
}
